### PR TITLE
[Rec-IM] Create & Edit custom participant in Admin page

### PR DESCRIPTION
**Pull request by Kevin-Chu58**
This pr adds a new form that allows creating custom participants, and updating their info in the Admin page.
Also improves the loading speed of Admin page by replacing `GetProfileInfo()` with a lighter method.

API pr: https://github.com/gordon-cs/gordon-360-api/pull/891.

New `Create Participant` Button
image

`Create Participant` View
image

Only `Custom Participant` has the `MoreOption` Button at the righthand side
And they don't have a pic for now
image

Menu displayed when clicked the `MoreOption` Button
image

`Edit Participant` View
image

Quick Search in RecIM team page
* Admins can search both Gordon participants and non-Gordon participants
image
image

**Participant List Checklist**
* let me know if I missed any
- [x] Admin page - participant list
- [x] Team page - roster
- [x] Team page - invite list
- [x] Match page - attendance list

### CODE REVIEW

File: src/services/recim/participant.ts
Comment (amos-cha) [inline src/services/recim/participant.ts:25]: Take a look at what we do for `activity`.ts, we can cut down on some of these repeated pieces of data. Also for the record @Kevin-Chu58's API PR MAKES a username that ensures no duplicates. However, I can see a small chance that there could be overlap if a Gordon User is added i.e:

silas.white (exists)
we add another silas.white -> silas.white2
Gordon adds another silas.white -> silas.white2 (conflict)

File: src/views/RecIM/components/Forms/ParticipantForm/index.jsx
Comment (amos-cha) [inline src/views/RecIM/components/Forms/ParticipantForm/index.jsx:59]: We have an "AllowEmails" but we don't have a way of storing emails of non-gordon participants. I think we either need another column for emails OR you need to set it always to false.

File: src/views/RecIM/components/List/Listing/index.jsx
Comment (amos-cha) [inline src/views/RecIM/components/List/Listing/index.jsx:273]:
```suggestion
        // TODO - Depends on how we want to store their Pics
```

**Review by amos-cha — CHANGES_REQUESTED** [review]
Small questions I need to have addressed. Besides that it looks good.

**Comment by amos-cha** [comment]
> Should we somehow change the wording to perhaps say "add a non-Gordon participant" or in some way clarify that this is only for making custom participants, not for adding someone to Rec-IM generally?

This isn't COMPLETELY necessary. If we were to fill in the exact data aka, `username`, `firstname`, `lastname` of a gordon student. It would still work perfectly fine since the only "gordon" feature is that we link to Gordon profiles

File: src/services/recim/participant.ts
Comment (Kevin-Chu58) [reply src/services/recim/participant.ts:25]: The autogenerated username algorithm is now fixed. We will always add a suffix `.custom` after each non-Gordon participant username.

File: src/views/RecIM/components/QuickSearch/index.jsx
Comment (amos-cha) [inline src/views/RecIM/components/QuickSearch/index.jsx:294]: hmmm i don't know if we should be ripping out the entire gordon search? I can't think of another solution off the top of my head but it feels a little sketchy?

Benefits are obviously that Rec-IM stays modular but this is definitely a repeated component?
thoughts @silasw2  @cpabbot

**Review by amos-cha — COMMENTED** [review]
I think we need a couple levels of approval for this. I want to keep Rec-IM modular but it feels weird to be ripping/copying exact components/functions from the rest of 360?

File: src/views/RecIM/components/QuickSearch/index.jsx
Comment (Kevin-Chu58) [reply src/views/RecIM/components/QuickSearch/index.jsx:294]: True. Most part is the same, except we also need to return `FirstName` and `LastName` otherwise the invite list shows `undefined undefined` for all people.
`    if (key === 'Enter' && suggestionList && suggestionList.length > 0) {
      if (sIndex === -1) {
        theChosenOne = suggestionList[0].UserName;
        firstName = suggestionList[0].firstName;
        lastName = suggestionList[0].lastName;
      } else {
        theChosenOne = suggestionList[sIndex].UserName;
        firstName = suggestionList[sIndex].firstName;
        lastName = suggestionList[sIndex].lastName;
      }
      // If prop set to disable link, then trigger the onSearchSubmit callback function
      // Else, redirect the user to the selected profile page
      disableLink
        ? onSearchSubmit(theChosenOne, firstName, lastName)`

Comment (amos-cha) [reply src/views/RecIM/components/QuickSearch/index.jsx:294]: Right, I agree that we have a unique piece of the component, but it is, in essence, identical. I'm just wondering if we should handle it this way or if we should modify the actual GordonQuickSearch which would interlink 360 and Rec-IM further

**Comment by Kevin-Chu58** [comment]
> This is looking great! One other quick note: can you change the gender dropdown to have full options "male, female, other/unspecified" (idk what to use for third but one of those maybe?) rather than just the letters?

image
@cpabbot Updated!

**Comment by amos-cha** [comment]
> > This is looking great! One other quick note: can you change the gender dropdown to have full options "male, female, other/unspecified" (idk what to use for third but one of those maybe?) rather than just the letters?
> 
> image @cpabbot Updated!

sharelock 🧑‍🍳

**Review by amos-cha — DISMISSED** [review]
I'm willing to approve, but wait for @silasw2 or @cpabbot to also approve the change.

**Comment by Kevin-Chu58** [comment]
> > > This is looking great! One other quick note: can you change the gender dropdown to have full options "male, female, other/unspecified" (idk what to use for third but one of those maybe?) rather than just the letters?
> > 
> > 
> > image @cpabbot Updated!
> 
> sharelock 🧑‍🍳

I am making a pun using the share lock and exclusive lock used to maintain the ACID of database, a daddy joke :)

File: src/views/RecIM/components/List/Listing/index.jsx
Comment (amos-cha) [inline src/views/RecIM/components/List/Listing/index.jsx:438]: This chunk of code seems odd. Why does a custom participant not need the `<ListItemButton>`. Also, if the nesting is required, can you just make a helper function that wraps the sub component? It's the exact same piece of code.

File: src/views/RecIM/components/List/Listing/index.jsx
Comment (amos-cha) [inline src/views/RecIM/components/List/Listing/index.jsx:462]: can we rename
```
moreOptionsCustomOpen
anchorCustomEl
```
to something else? Those variable names are not very descriptive. moreParticipantOptions? participantAnchorEl?

File: src/views/RecIM/components/List/Listing/index.jsx
Comment (amos-cha) [inline src/views/RecIM/components/List/Listing/index.jsx:475]:
```suggestion
                editDetails(participant);
```
editDetails of what? descriptive?

**Review by amos-cha — CHANGES_REQUESTED** [review]
Few changes to be addressed

**Review by amos-cha — APPROVED** [review]
Forced to approve. API now fails without new UI

**Pull request merged**
